service/codegen/src/main/kotlin/evaka/codegen/apitypes/Generate.kt

```diff
@@ -188,16 +169,70 @@ ${values.joinToString("\n") { "  | '$it'" }}"""
     }
 }
 
-private data class AnalyzedProperty(
+data class AnalyzedProperty(
     val name: String,
-    val type: KClass<*>,
-    val nullable: Boolean,
-    val collection: Boolean
+    val type: AnalyzedType
 ) {
-    fun toTs(): String {
-        return "$name: ${tsMapping[type.qualifiedName]?.type ?: type.simpleName}"
-            .let { if (collection) "$it[]" else it }
-            .let { if (nullable) "$it | null" else it }
+    fun toTs() = "$name: ${type.toTs()}"
+}
+
+fun analyzeType(type: KType): AnalyzedType = when {
+    isMap(type) -> TsMap(type)
+    isCollection(type) -> TsArray(type)
+    else -> TsPlain(type)
+}
+
+sealed interface AnalyzedType {
+    val declarableTypes: List<KClass<*>>
+    fun toTs(): String
+}
+
+data class TsPlain(val type: KType) : AnalyzedType {
+    override val declarableTypes = listOf(type.jvmErasure)
+    override fun toTs() = toNullableTs(type.isMarkedNullable, tsName(type.jvmErasure))
+}
+
+data class TsArray(val type: KType) : AnalyzedType {
+    private val typeParameter = unwrapCollection(type)
+    override val declarableTypes = listOf(typeParameter)
+    override fun toTs() = toNullableTs(type.isMarkedNullable, "Array<${tsName(typeParameter)}>")
```

Comment:
Why did change `X[]` to `Array<X>`? The diff is unnecessarily large and we use `X[]` in hand-crafted code, too.

Reply:
It might not be a very relevant concern, but collections of nullable types like `List<T?>` would need extra parentheses `(T | null)[]` but `Array<T | null>` works automatically.

Reply:
Arrays are now written as `X[]`. Nullable types weren't handled correctly anyway so I had to fix it separately.